eqcorrscan/core/match_filter.py

```diff
@@ -3360,6 +3386,170 @@ def normxcorr2(template, image):
+    :param stream:
+        A single Stream object to be correlated with the templates.  This is
+        in effect the image in normxcorr2 and cv2.
+    :type dask: bool
+    :param dask:
+        Whether to use dask for multiprocessing or not, if False, will use
+        python native multiprocessing.
+    :type compute: bool
+    :param compute:
+        Only valid if dask==True. If compute==False, returned result with be
+        a dask.delayed object, useful if you are using dask to compute multiple
+        time-steps at the same time.
+    :type cores: int
+    :param cores:
+        Number of processed to use, if set to None, and dask==False, no
+        multiprocessing will be done.
+    :type cores: int
+    :param cores: Number of cores to loop over
+
+    :returns:
+        New list of :class:`numpy.ndarray` objects.  These will contain
+        the correlation sums for each template for this day of data.
+    :rtype: list
+    :returns:
+        list of ints as number of channels used for each cross-correlation.
+    :rtype: list
+    :returns:
+        list of list of tuples of station, channel for all cross-correlations.
+    :rtype: list
+
+    .. Note::
+        Each template must contain the same channels as every other template,
+        the stream must also contain the same channels (note that if there
+        are duplicate channels in the template you do not need duplicate
+        channels in the stream).
+    """
+    no_chans = np.zeros(len(templates))
+    chans = [[] for _i in range(len(templates))]
+    # Do some reshaping
+    stream.sort(['network', 'station', 'location', 'channel'])
+    t_starts = []
+    for template in templates:
+        template.sort(['network', 'station', 'location', 'channel'])
+        t_starts.append(min([tr.stats.starttime for tr in template]))
+    seed_ids = [tr.id + '_' + str(i) for i, tr in enumerate(templates[0])]
+    template_array = {}
+    stream_array = {}
+    pad_array = {}
+    for i, seed_id in enumerate(seed_ids):
+        t_ar = np.array([template[i].data for template in templates])
+        template_array.update({seed_id: t_ar})
+        stream_array.update(
+            {seed_id: stream.select(id=seed_id.split('_')[0])[0].data})
+        pad_list = [
+            int(round(template[i].stats.sampling_rate *
+                      (template[i].stats.starttime - t_starts[j])))
+            for j, template in zip(range(len(templates)), templates)]
+        pad_array.update({seed_id: pad_list})
+    # if use_dask:
+    #     import dask
+    #     xcorrs = []
+    #     for seed_id in seed_ids:
+    #         tr_xcorrs, tr_chans = dask.delayed(multi_normxcorr)(
+    #             templates=template_array[seed_id],
+    #             stream=stream.select(id=seed_id.split('_')[0])[0].data)
+    #         xcorrs.append(tr_xcorrs)
+    #     cccsums = dask.delayed(np.sum)(xcorrs, axis=0)
+    #     if compute:
+    #         cccsums.compute()
+    if cores is None:
```

Comment:
Would it make sense to abstract the multiprocessing further up? I know several functions use something similar so maybe we could make a generic pool interface on the module level, then we could have persistent processes/threads in the pool so we wouldn't need to spin them up every time.